### ISSUE TITLE
Multiple language server instances on save

### ISSUE THREAD

**Issue by eratio08** [nvim-lua/kickstart.nvim#4]
Not sure if this is intended but saving the `init.lua` file will cause a new instance of `sumneko` getting started. I assuming that this is not intended as it will slow down vim eventually.

**Comment by skovati** [comment]
Strange, on my machine it just respawns `lua-language-server` but never has more than one instance running at a time. Are you getting multiple instances running concurrently? Seems like an easy enough solution would be to remove [this](https://github.com/nvim-lua/kickstart.nvim/blob/master/init.lua#L54) `source` that runs every time we save `init.lua`.

**Comment by gugahoi** [comment]
I believe the problems stems from [L53](https://github.com/nvim-lua/kickstart.nvim/blob/5788a41cb82e3598fccbe99f4661e8372ff59854/init.lua#L53)

```
-- Automatically source and re-compile packer whenever you save this init.lua
local packer_group = vim.api.nvim_create_augroup('Packer', { clear = true })
vim.api.nvim_create_autocmd('BufWritePost', {
  command = 'source <afile> | PackerCompile',
  group = packer_group,
  pattern = vim.fn.expand '$MYVIMRC',
})
```

**Comment by eratio08** [comment]
> Strange, on my machine it just respawns lua-language-server but never has more than one instance running at a time. Are you getting multiple instances running concurrently? Seems like an easy enough solution would be to remove this source that runs every time we save init.lua.

Locking at `LspInfo` I would say there are multiple instances. Saving also uses increasingly more cpu each consequent time
![screenshot_2022-06-26_18-05-58_679704351](https://user-images.githubusercontent.com/5063643/175823263-e87c1c71-80af-4b38-afc5-bb060f72707d.png)

**Comment by eratio08** [comment]
> 



> I believe the problem stems from [L53](https://github.com/nvim-lua/kickstart.nvim/blob/5788a41cb82e3598fccbe99f4661e8372ff59854/init.lua#L53)

Yes, I know but getting rid of it will also stop auto-sourcing. I was hoping that there might be a trick to have it lazy init the server.

**Comment by eratio08** [comment]
Fix for another config was to comment out `workspace = { library = vim.api.nvim_get_runtime_file('', true) },` before that it would be very slowly starting server and it was starting always 2 server at the same time.

**Comment by anthony-S93** [comment]
@eratio08 

I myself stumbled upon this issue today. Turning off single file support does prevent multiple instances of `sumneko_lua` from being attached to the file after the sourcing in the auto-command; however, what if we need `single_file_support`? 

To solve this problem, I wrote a few lines of code that sort of wrap around your `require('lspconfig').sumneko_lua.setup{}` logic. 

Note that this assumes you are calling the `setup{}` function in the same file containing the auto-command. For me, that's inside the `plugins.lua` file: 

```lua
vim.cmd([[
      augroup packer_user_config
          autocmd!
          autocmd BufWritePost plugins.lua source <afile> | PackerCompile
      augroup end
  ]])

...Your other config...

--Setting up language servers
local lspconfig = require("lspconfig")
local start_sumneko_lua = true --boolean defaults to true, otherwise, the server will never start
local current_buf_id = vim.api.nvim_get_current_buf()
local servers_attached_to_current_buf = vim.lsp.get_active_clients({bufnr = current_buf_id})

for _, server in ipairs(servers_attached_to_current_buf) do
    if server.name == "sumneko_lua" then --an instance of sumneko_lua is already attached to the buffer
        start_sumneko_lua = false 
    end
end

--After that, all you have to do is to wrap sumneko lua's setup() call inside a conditional using the boolean

if start_sumneko_lua then
    lspconfig.sumneko_lua.setup{
         -- yada yada yada
    }
end

--Other servers don't need this, so their setup can be called as usual
lspconfig.pyright.setup{

}

--yada yada yada...

```
The idea is simply to use some kind of "flag" to determine whether or not to call the setup() function for sumneko_lua (since the setup function determines the availability of the language server). The flag defaults to `true` but will be switched to `false` if the current buffer already has an instance of `sumneko_lua` attached to it. 

Granted, the code is a bit inelegant and 'hackish', but it worked without any issues for me because all my plugin configuration is inside packer.nvim's plugin specification file (`plugins.lua`). 

I hope this helps.

**Comment by VonHeikemen** [comment]
Problem happens when you source the file more than once. I would just put all the language servers inside an `if` that checks the value of a global variable.

```lua
if vim.g.user_lsp_setup == nil then
  vim.g.user_lsp_setup = true

  require('lspconfig').sumneko_lua.setup({
    on_attach = on_attach,
    capabilities = capabilities,
    --- more code.....
  })

  require('lspconfig').tsserver.setup({
    on_attach = on_attach,
    capabilities = capabilities,
  })
end
```
